backpack/core/derivatives/lstm.py

```diff
@@ -0,0 +1,358 @@
+        for t in range(T):
+            if t == 0:
+                ifgo[t] = (
+                    einsum("hi, ni -> nh", module.weight_ih_l0, module.input0[t])
+                    + module.bias_ih_l0
+                    + module.bias_hh_l0
+                )
+            else:
+                ifgo[t] = (
+                    einsum("hi, ni -> nh", module.weight_ih_l0, module.input0[t])
+                    + module.bias_ih_l0
+                    + einsum("hg, ng -> nh", module.weight_hh_l0, module.output[t - 1])
+                    + module.bias_hh_l0
+                )
+            ifgo[t, :, H0:H1] = sigmoid(ifgo[t, :, H0:H1])
+            ifgo[t, :, H1:H2] = sigmoid(ifgo[t, :, H1:H2])
+            ifgo[t, :, H2:H3] = tanh(ifgo[t, :, H2:H3])
+            ifgo[t, :, H3:H4] = sigmoid(ifgo[t, :, H3:H4])
+            if t == 0:
+                c[t] = ifgo[t, :, :H] * ifgo[t, :, H2:H3]
+            else:
+                c[t] = (ifgo[t, :, :H] * ifgo[t, :, H2:H3]) + (
+                    ifgo[t, :, H:H2] * c[t - 1]
+                )
+            c_tanh[t] = tanh(c[t])
+            h[t] = ifgo[t, :, H3:] * c_tanh[t]
+
+        # h is the same as previous forward pass
+        assert allclose(h, module.output, atol=1e-5)
+        return ifgo, c, c_tanh, h
+
+    @classmethod
+    def _ifgo_jac_t_mat_prod(cls, module: LSTM, mat: Tensor) -> Tensor:
+        V: int = mat.shape[0]
+        T: int = mat.shape[1]
+        N: int = mat.shape[2]
+        H: int = module.hidden_size
+        H0: int = 0 * H
+        H1: int = 1 * H
+        H2: int = 2 * H
+        H3: int = 3 * H
+        H4: int = 4 * H
+
+        ifgo, c, c_tanh, _ = cls._forward_pass(module, mat)
+
+        # backward pass
+        H_prod: Tensor = zeros(V, T, N, H, device=mat.device)
+        C_prod: Tensor = zeros(V, T, N, H, device=mat.device)
+        IFGO_prod: Tensor = zeros(V, T, N, 4 * H, device=mat.device)
+        for t in range(T)[::-1]:
+            # jac_t_mat_prod until node h
+            if t == T - 1:
+                H_prod[:, t] = mat[:, t]
+            else:
+                H_prod[:, t] = mat[:, t] + einsum(
+                    "vnh, hg -> vng",
+                    IFGO_prod[:, t + 1],
+                    module.weight_hh_l0,
+                )
+
+            # C_prod = jac_t_mat_prod until node c
+            if t == T - 1:
+                C_prod[:, t] = einsum(
+                    "vnh, nh, nh -> vnh",
+                    H_prod[:, t],
+                    ifgo[t, :, H3:H4],
+                    (1 - c_tanh[t] ** 2),
+                )
+            else:
+                C_prod[:, t] = einsum(
+                    "vnh, nh, nh -> vnh",
+                    H_prod[:, t],
+                    ifgo[t, :, H3:H4],
+                    (1 - c_tanh[t] ** 2),
+                ) + einsum(
+                    "vnh, nh -> vnh",
+                    C_prod[:, t + 1],
+                    ifgo[t + 1, :, H1:H2],
+                )
```

Comment:
Can you try a similar refactoring as proposed above to avoid duplicating the terms that appear in both of the `if/else` branches?

Reply:
Yes, except in-place (see discussion above).